Add docs for CRD field selection (#45155)
kubernetes/website: content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions.md

Reviewer: The problem with adding it to a task page is: what does the task title end up as?
“Extend Kubernetes using a CustomResourceDefinition” is one thing; if the task is “Try out lots of different ways to use CustomResourceDefinitions”, we call that a tutorial. Tasks should be specific and the title should match what users might actually do; for example, “Install kubectl” or “Get a Shell to a Running Container”.

Author: I agree.

We have a bit of a mess..  For historical reasons, CRDs somehow got almost completely documented here in the tasks page, which is the wrong place for almost all of the documentation.  We probably should have almost all this documentation in https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/ or somewhere more canonical (suggestions welcome on where exactly that should be).

I biased toward adding this enhancmenet's docs adjacent to the docs for similar fields like additional printer columns.

I think we do need to fix this.  Should we do it as part of this enhancements documentation? I feel like it will be a larger project..  I'd admit I've been putting it off.

Reviewer: What's the smallest change we can make to the concept page to allow documenting this feature in the place where it belongs?

Let's do that and then see if we have capacity to iterate.

Reviewer: https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/ is a good home for explaining CRDs and their details (tutorials aside).

Author: SG. I've moved the docs into concepts. I've simplified and rephrased the docs slightly to have more of a explanatory approach (and less of a tutorial approach).

The `NAME` column is implicit and does not need to be defined in the CustomResourceDefinition.
{{< /note >}}

### Field Selectors

[Field Selectors](/docs/concepts/overview/working-with-objects/field-selectors/)
let clients select custom resources based on the value of one or more resource
fields.

All custom resources support the `metadata.name` and `metadata.namespace` field
selectors.

Fields declared in a `CustomResourceDefinition` may also be used with field
selectors when `selectableFields` is used.

> #### SelectableFields

Reviewer: Specifying the anchor string means we can confidently change the heading without breaking existing bookmarks.

use this behavior, which then applies to all CustomResourceDefinitions in your
cluster.

The `selectableFields` field of `CustomResourceDefinition` may be used to
declare which other fields in a custom resource may be used in field selectors.
The following example adds the `.spec.color` and `.spec.size` fields as
selectable fields.

Save the CustomResourceDefinition to `shirt-resource-definition.yaml`:

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: shirt.stable.example.com
spec:
  group: stable.example.com
  scope: Namespaced
  names:
    plural: shirts
    singular: shirt
    kind: Shirt
  versions:
  - name: v1
    served: true
    storage: true
    schema:
      openAPIV3Schema:
        type: object
        properties:
          spec:
            type: object
            properties:
              color:
                type: string
              size:
                type: string
      selectableFields:
      - jsonPath: .spec.color
      - jsonPath: .spec.size
      additionalPrinterColumns:
      - jsonPath: .spec.color
        name: Color
        type: string
      - jsonPath: .spec.size
        name: Size

>         type: string

Reviewer: We'd prefer a downloadable example manifest using `{{% code_sample %}}` shortcode; however, this is OK as-is for alpha.

Author: Good call. I'll switch to this now.

```

Create the CustomResourceDefinition:

```shell

> kubectl apply -f shirt-resource-definition.yaml

Reviewer: If it were downloadable, you could specify the URL here.

Reviewer: ```suggestion
kubectl apply -f https://k8s.io/examples/examples/customresourcedefinition/shirt-resource-definition.yaml
```

Reviewer: (and no need to save it first)

```

Save some custom resources to `shirt-resources.yaml`:

```yaml

> apiVersion: stable.example.com/v1

Reviewer: ```suggestion
---
apiVersion: stable.example.com/v1
```

When there's more than one document, I like to start each YAML document with a marker line (even the first one).

kind: CustomResourceDefinition
metadata:
  name: example1
spec:
  color: blue
  size: S
---
apiVersion: stable.example.com/v1
kind: CustomResourceDefinition
metadata:
  name: example2
spec:
  color: blue
  size: M
---
apiVersion: stable.example.com/v1
kind: CustomResourceDefinition
metadata:
  name: example2
spec:
  color: green
  size: M
```

Create the custom resources:

```shell
kubectl apply -f shirt-resources.yaml
```

Get all the resources:

```shell
$ kubectl get selector.stable.example.com
NAME       COLOR  SIZE
example1   blue   S
example2   blue   M

> example3   green  M

Reviewer: We [don't include the prompt](https://kubernetes.io/docs/contribute/style/style-guide/#don-t-include-the-command-prompt).

```

Get only resources with with a `color` of `blue`:

```shell

> $ kubectl get selector.stable.example.com --field-selector spec.color=blue

Reviewer: We [don't include the prompt](https://kubernetes.io/docs/contribute/style/style-guide/#don-t-include-the-command-prompt).

NAME       COLOR  SIZE
example1   blue   S
example2   blue   M
```

Get only resources with a `color` of `green` and a `size` of `M`:

```shell

> $ kubectl get selector.stable.example.com --field-selector spec.color=green,spec.size=M

Reviewer: We [don't include the prompt](https://kubernetes.io/docs/contribute/style/style-guide/#don-t-include-the-command-prompt).

let clients select custom resources based on the value of one or more resource
fields.

All custom resources support the `metadata.name` and `metadata.namespace` field
selectors.

Fields declared in a {{< glossary_tooltip term_id="CustomResourceDefinition" text="CustomResourceDefinition" >}}
may also be used with field
selectors when `selectableFields` is used.

#### Selectable fields for custom resources {#crd-selectable-fields}

{{< feature-state state="alpha" for_k8s_version="v1.30" >}}

You need to enable the `CustomResourceFieldSelectors`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) to
use this behavior, which then applies to all CustomResourceDefinitions in your
cluster.

The `selectableFields` field of `CustomResourceDefinition` may be used to
declare which other fields in a custom resource may be used in field selectors.
The following example adds the `.spec.color` and `.spec.size` fields as
selectable fields.

Save the CustomResourceDefinition to `shirt-resource-definition.yaml`:

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: shirt.stable.example.com
spec:

>   group: stable.example.com

Reviewer: ```suggestion
  group: stable.example.com # an example; use your own API group for real APIs
```

  scope: Namespaced
  names:
    plural: shirts
    singular: shirt
    kind: Shirt
  versions:
  - name: v1
    served: true
    storage: true
    schema:
      openAPIV3Schema:
        type: object
        properties:
          spec:
            type: object
            properties:
              color:
                type: string
              size:
                type: string
      selectableFields:
      - jsonPath: .spec.color
      - jsonPath: .spec.size
      additionalPrinterColumns:
      - jsonPath: .spec.color
        name: Color
        type: string
      - jsonPath: .spec.size
        name: Size
        type: string
```

Create the CustomResourceDefinition:

```shell
kubectl apply -f https://k8s.io/examples/examples/customresourcedefinition/shirt-resource-definition.yaml
```

> Save some custom resources to `shirt-resources.yaml`:

Reviewer: ```suggestion
Define some Shirts by editing `shirt-resources.yaml`; for example:
```

```yaml
---
apiVersion: stable.example.com/v1

> kind: CustomResourceDefinition

Reviewer: Did you mean:
```suggestion
kind: Shirt
```

metadata:
  name: example1
spec:
  color: blue
  size: S
---
apiVersion: stable.example.com/v1

> kind: CustomResourceDefinition

Reviewer: Did you mean:
```suggestion
kind: Shirt
```

metadata:
  name: example2
spec:
  color: blue
  size: M
---
apiVersion: stable.example.com/v1

> kind: CustomResourceDefinition

Reviewer: Did you mean:
```suggestion
kind: Shirt
```

metadata:
  name: example2
spec:
  color: green
  size: M
```

Create the custom resources:

```shell
kubectl apply -f shirt-resources.yaml
```

Get all the resources:

```shell

> kubectl get selector.stable.example.com

Reviewer: Is this right?

Reviewer: One more:
```suggestion
kubectl get shirts.stable.example.com
```

```

> Should output:

Reviewer: (style nit)
```suggestion
The output is:
```

```
NAME       COLOR  SIZE
example1   blue   S
example2   blue   M
example3   green  M
```

> Get only resources with with a `color` of `blue`:

Reviewer: ```suggestion
Fetch blue shirts (retrieve Shirts with a `color` of `blue`):
```

```shell

> kubectl get selector.stable.example.com --field-selector spec.color=blue

Reviewer: ```suggestion
kubectl get shirts.stable.example.com --field-selector spec.color=blue
```
Or:
```suggestion
kubectl get shirts --field-selector spec.color=blue
```

```

Should output:

```
NAME       COLOR  SIZE
example1   blue   S
example2   blue   M
```

Get only resources with a `color` of `green` and a `size` of `M`:

```shell

> kubectl get selector.stable.example.com --field-selector spec.color=green,spec.size=M

Reviewer: ```suggestion
kubectl get shirts.stable.example.com --field-selector spec.color=green,spec.size=M
```

The `NAME` column is implicit and does not need to be defined in the CustomResourceDefinition.
{{< /note >}}

> ### Field Selectors

Reviewer: (nit)
```suggestion
### Field selectors
```
